File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.java
class Solution {
    private void dfs(List<String> results, Map<String, PriorityQueue<String>> fromToMap, String from) {
        while (fromToMap.containsKey(from) && !fromToMap.get(from).isEmpty()) {
            dfs(results, fromToMap, fromToMap.get(from).poll());
        }

        results.add(0, from);
    }

    public List<String> findItinerary(List<List<String>> tickets) {
        List<String> results = new LinkedList<>();
        Map<String, PriorityQueue<String>> fromToMap = new HashMap<>();

        for (List<String> ticket : tickets) {
            fromToMap.putIfAbsent(ticket.get(0), new PriorityQueue<>());
            fromToMap.get(ticket.get(0)).add(ticket.get(1));
        }

        dfs(results, fromToMap, "JFK");

        return results;
    }
}

File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.kt
class Solution {
    fun findItinerary(tickets: List<List<String>>): List<String> {
        val results = mutableListOf<String>()
        val fromToMap = mutableMapOf<String, PriorityQueue<String>>()

        tickets.forEach { (from, to) ->
            fromToMap.getOrPut(from) { PriorityQueue() }.add(to)
        }

        fun dfs(from: String) {
            val destinations = fromToMap[from]
            while (destinations != null && destinations.isNotEmpty()) {
                dfs(destinations.poll())
            }
            results.add(0, from)
        }

        dfs("JFK")

        return results
    }
}